components/script/dom/readablestream.rs

```diff
@@ -0,0 +1,517 @@
+            js_stream: Heap::default(),
+            js_reader: Heap::default(),
+            has_reader: Default::default(),
+            external_underlying_source: external_underlying_source,
+        }
+    }
+
+    fn new(
+        global: &GlobalScope,
+        external_underlying_source: Option<Rc<ExternalUnderlyingSourceController>>,
+    ) -> DomRoot<ReadableStream> {
+        reflect_dom_object(
+            Box::new(ReadableStream::new_inherited(external_underlying_source)),
+            global,
+        )
+    }
+
+    /// Used from RustCodegen.py
+    #[allow(unsafe_code)]
+    pub fn from_js(cx: SafeJSContext, obj: *mut JSObject) -> Result<DomRoot<ReadableStream>, ()> {
+        unsafe {
+            if !IsReadableStream(obj) {
+                return Err(());
+            }
+
+            let in_realm_proof = AlreadyInRealm::assert_for_cx(cx);
+            let global = GlobalScope::from_safe_context(cx, InRealm::Already(&in_realm_proof));
+
+            let stream = ReadableStream::new(&global, None);
+            stream.js_stream.set(UnwrapReadableStream(obj));
+
+            Ok(stream)
+        }
+    }
+
+    /// Build a stream backed by a Rust underlying source.
+    #[allow(unsafe_code)]
+    pub fn new_with_external_underlying_source(
+        global: &GlobalScope,
+        source: ExternalUnderlyingSource,
+    ) -> DomRoot<ReadableStream> {
+        let _ar = enter_realm(global);
+        let cx = global.get_cx();
+
+        let source = Rc::new(ExternalUnderlyingSourceController::new(source));
+
+        let stream = ReadableStream::new(&global, Some(source.clone()));
+
+        unsafe {
+            let mut traps = ReadableStreamUnderlyingSourceTraps {
+                requestData: Some(request_data),
+                writeIntoReadRequestBuffer: Some(write_into_read_request_buffer),
+                cancel: Some(cancel),
+                onClosed: Some(close),
+                onErrored: Some(error),
+                finalize: Some(finalize),
+            };
+
+            let js_wrapper = CreateReadableStreamUnderlyingSource(
+                &mut traps,
+                &*source as *const _ as *const c_void,
+            );
+
+            rooted!(in(*cx)
+                let js_stream = NewReadableExternalSourceStreamObject(
+                    *cx,
+                    js_wrapper,
+                    ptr::null_mut() as *mut c_void,
+                    HandleObject::null(),
+                )
+            );
+
+            stream.js_stream.set(UnwrapReadableStream(js_stream.get()));
+        }
+
+        stream
+    }
+
+    /// Get a pointer to the underlying JS object.
+    pub fn get_js_stream(&self) -> NonNull<JSObject> {
+        NonNull::new(self.js_stream.get())
+            .expect("Couldn't get a non-null pointer to JS stream object.")
+    }
+
+    #[allow(unsafe_code)]
+    pub fn enqueue_native(&self, bytes: Vec<u8>) {
+        let global = self.global();
+        let _ar = enter_realm(&*global);
```

Comment:
I think it's necessary because when you enqueue bytes via the JS_ api, this ends-up resolving the read promise(if there is any outstanding). 

Yes I'll switch to add the proof as an argument.